src/screens/history/HistoryScreen.tsx

```diff
@@ -43,12 +43,23 @@ const HistoryScreen = ({ navigation }: HistoryScreenProps) => {
     const dateGroups = rawHistory.reduce<Record<string, History[]>>(
       (groups, item) => {
         const date = convertDateToISOString(item.readTime);
-
         if (!groups[date]) {
           groups[date] = [];
         }
-
-        groups[date].push(item);
+        const lastNovel = groups[date][groups[date].length - 1];
+
+        if (
+          (lastNovel?.novelId === item.novelId && lastNovel.startChapter) ??
+          lastNovel.chapterNumber === (item.chapterNumber ?? 0) + 1
+        ) {
+          groups[date][groups[date].length - 1] = {
+            ...lastNovel,
+            endChapter: lastNovel.endChapter ?? lastNovel.chapterNumber,
+            startChapter: item.chapterNumber,
+          };
+        } else {
+          groups[date].push(item);
+        }
```

Comment:
It could be that the logic breaks in some cases.
First case: It won't since the history array would store it as you wrote, which means that the ```lastNovel?.novelId === item.novelId``` check would fail and restart the compacting process on the next novel.
Second case: Why would someone do this
Third case: Could be, I don't know a novel with decimal chapter numbers and don't have time currently to search for one. However I assuemend the ```.chapterNumber``` would be added by the plugin and only contain a whole number. Correct me if I'm wrong.

The ?? is at the right place, however I somehow misplaced the Parentheses.